fiftyone/utils/youtubedl.py

```diff
@@ -0,0 +1,343 @@
+"""
+Utilities for working with
+`youtube-dl <https://github.com/ytdl-org/youtube-dl>`
+
+| Copyright 2017-2021, Voxel51, Inc.
+| `voxel51.com <https://voxel51.com/>`_
+|
+"""
+from collections import defaultdict
+import logging
+import multiprocessing
+import multiprocessing.dummy
+import os
+
+import eta.core.utils as etau
+import eta.core.video as etav
+
+import fiftyone.core.utils as fou
+
+youtube_dl = fou.lazy_import(
+    "youtube_dl",
+    callback=lambda: fou.ensure_import(
+        "youtube_dl",
+        error_msg="youtube_dl not found, run `pip install youtube-dl`",
+    ),
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+def download_from_youtube(
+    urls,
+    download_dir=None,
+    clip_segments=None,
+    max_videos=None,
+    num_workers=None,
+    ext=None,
+    buffer_seconds=5,
+):
+    """
+    Attempts to download a list of video urls from YouTube.
+    The `urls` argument either accepts:
+
+        * A list of YouTube video urls::
+
+            urls = ["https://www.youtube.com/watch?v=-0URMJE8_PB", ...]
+
+          When `urls` is a list, then the `download_dir` argument is required
+          and all videos will be downloaded into that directory
+
+        * A dictionary mapping the video urls to files on disk in which to store that video::
+
+            urls = {
+                "https://www.youtube.com/watch?v=-0URMJE8_PB": "/path/to/local/file1.ext",
+                ...
+            }
+
+    The corresponding `clip_segments` argument can then contain a list oftuples of ints
+    or floats used to download segment clips of each video. A value of `None`
+    can indicate either downloading the entire video, downloading from the
+    start or downloading to the end of the video::
+
+        clip_segments = [
+            (10, 25),
+            (11.1, 20.2),
+            None,
+            (None, 8.3),
+        ]
+
+    Args:
+        urls: either a list of video urls, or a dict
+            mapping these urls to locations on disk. If `urls` is a list, then
+            the `download_dir` argument is required
+        download_dir (None): the output directory used to store the downloaded
+            videos. This is only used if `urls` is a list
+        clip_segments (None): a list of int or float tuples indicating start
```

Comment:
Things will be undefined when `urls` is a dict but `clip_segments` is a list, right?

Reply:
I guess the best solution here is to have `urls`, `video_paths`, and `clip_segments` arguments that are all lists. 

This is specifically for clip downloads in a `download_youtube_clips()` method. For full video downloads in a `download_youtube_videos()` method, you could still allow `urls` to be a dict if you want.